metagpt/actions/research.py

```diff
@@ -0,0 +1,288 @@
+        self.search_engine = SearchEngine()
+        self.rank_func = rank_func
+
+    async def run(
+        self,
+        topic: str,
+        decomposition_nums: int = 4,
+        url_per_query: int = 4,
+        system_text: str | None = None,
+    ) -> dict[str, list[str]]:
+        """Run the action to collect links.
+
+        Args:
+            topic: The research topic.
+            decomposition_nums: The number of search questions to generate.
+            url_per_query: The number of URLs to collect per search question.
+            system_text: The system text.
+
+        Returns:
+            A dictionary containing the search questions as keys and the collected URLs as values.
+        """
+        system_text = system_text if system_text else RESEARCH_TOPIC_SYSTEM.format(topic=topic)
+        search_topic_prompt = SEARCH_TOPIC_PROMPT.format(topic=topic)
+        logger.debug(search_topic_prompt)
+        keywords = await self._aask(search_topic_prompt, [system_text])
+        try:
+            keywords = json.loads(keywords)
+            keywords = parse_obj_as(list[str], keywords)
+        except Exception as e:
+            logger.exception(f"fail to get keywords related to the research topic \"{topic}\" for {e}")
+            keywords = [topic]
+        results = await asyncio.gather(*(self.search_engine.run(i, as_string=False) for i in keywords))
+
+        def gen_msg():
+            while True:
+                search = "\n".join(f"#### Keyword: {i}\n Search Result: {j}\n" for (i, j) in zip(keywords, results))
+                prompt = SUMMARIZE_SEARCH_PROMPT.format(decomposition_nums=decomposition_nums, search=search)
+                yield prompt
+                remove = max(results, key=len)
+                remove.pop()
+                if len(remove) == 0:
+                    break
+        prompt = reduce_message_length(gen_msg(), self.llm.model, system_text, CONFIG.max_tokens_rsp)
+        logger.debug(prompt)
+        queries = await self._aask(prompt, [system_text])
+        try:
+            queries = json.loads(queries)
+            queries = parse_obj_as(list[str], queries)
+        except Exception as e:
+            logger.exception(f"fail to break down the research question for {e}")
+            queries = keywords
+        ret = {}
+        for query in queries:
+            ret[query] = await self._search_and_rank_urls(topic, query, url_per_query)
+        return ret
+
+    async def _search_and_rank_urls(self, topic: str, query: str, num_results: int = 4) -> list[str]:
+        """Search and rank URLs based on a query.
+
+        Args:
+            topic: The research topic.
+            query: The search query.
+            num_results: The number of URLs to collect.
+
+        Returns:
+            A list of ranked URLs.
+        """
+        max_results = max(num_results * 2, 6)
+        results = await self.search_engine.run(query, max_results=max_results, as_string=False)
```

Comment:
I appreciate your suggestion to reuse the keyword search results in case of query parsing failure. While it's true that reusing keyword search results might enhance the overall robustness, it's important to consider that the occurrence of query parsing failure is relatively low. Given this, the potential benefits of introducing such complexity might not outweigh the gains, especially considering the slight differences in input between the two scenarios.

Reply:
Conducting two separate searches serves a specific purpose in enhancing the quality and effectiveness of the interactions with the language model. The decision to perform these two searches is based on the need to provide contextual information and background to the language model, particularly when encountering new terms or concepts that the LLM might not have precise knowledge of.

By first performing a keyword search and collecting results, you're effectively furnishing the LLM with the necessary background and context to better understand the topic and the potential sub-questions that need to be addressed. This preliminary step serves as a way to prepare the LLM for the subsequent breakdown of the research question into sub-problems.